linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala

```diff
@@ -1,121 +1,427 @@
+        s"${d.timestamp.inNanoseconds} ${d.deadline.inNanoseconds}"
+
+      def set(headers: HeaderMap, deadline: FDeadline): Unit = {
+        val _ = headers.set(Key, write(deadline))
+      }
+
+      def clear(headers: HeaderMap): Unit = {
+        val _ = headers.remove(Key)
+      }
+
+      /**
+       * Extract the deadline from the request and, if it exists, use
+       * either the strictest combination of deadlines.
+       *
+       * Clears deadline headers from the request. This means that the
+       * client is responsible for encoding outgoing deadlines.
+       */
+      class ServerFilter extends SimpleFilter[Request, Response] {
+        def apply(req: Request, service: Service[Request, Response]) =
+          get(req.headerMap) match {
+            case None => service(req)
+            case Some(reqDeadline) =>
+              clear(req.headerMap)
+              val deadline = FDeadline.current match {
+                case None => reqDeadline
+                case Some(current) => FDeadline.combined(reqDeadline, current)
+              }
+              Contexts.broadcast.let(FDeadline, deadline) {
+                service(req)
+              }
+          }
+      }
+
+      /**
+       * If a deadline is set, encode it on downstream requests.
+       *
+       * Clears any existing deadline headers from the request.
+       */
+      class ClientFilter extends SimpleFilter[Request, Response] {
+        def apply(req: Request, service: Service[Request, Response]) =
+          FDeadline.current match {
+            case None => service(req)
+            case Some(deadline) =>
+              set(req.headerMap, deadline)
+              service(req)
+          }
+      }
     }
 
-    def clear(headers: HeaderMap): Unit = {
-      val _ = headers.remove(Key)
+    /**
+     * There are two headers used to control local Dtabs in linkerd:
+     *
+     *   1. `l5d-ctx-dtab` is read and _written_ by linkerd. It is
+     *      intended to managed entirely by linked, and applications
+     *      should only forward requests prefixed by `l5d-ctx-*`.
+     *
+     *      *NOTE*: the client module does not yet encode
+     *      `l5d-ctx-dtab`. `dtab-local` is still to be relied on
+     *      until https://github.com/twitter/finagle/pull/514 is
+     *      complete.
+     *
+     *   2. `l5d-dtab` is to be provided by users. Applications are
+     *       not required to forward `l5d-dtab` when fronted by
+     *       linkerd.
+     *
+     * `l5d-dtab` is appended to `l5d-ctx-dtab`, so that user-provided
+     * delegations take precdence.
+     */
+    object Dtab {
+      val CtxKey = Ctx.Prefix + "dtab"
+      val UserKey = Headers.Prefix + "dtab"
+
+      private val EmptyReturn = Return(FDtab.empty)
+
+      def get(headers: HeaderMap, key: String): Try[FDtab] =
+        if (!headers.contains(key)) EmptyReturn
+        else Try { FDtab(headers.getAll(key).view.flatMap(FDtab.read(_)).toIndexedSeq) }
```

Comment:
I assume the `view` call here is to save on allocating the intermediate collection? If so, IIRC using `breakOut` would be even cheaper, but I haven't done the microbenchmark recently:
```scala
else Try { FDtab(headers.getAll(key).flatMap(FDtab.read)(breakOut)) }
```